### PR TITLE
applications: nrf_desktop: Add peers cache to HID forward module

### PR DESCRIPTION
**Pull request by MarekPieta**
PR introduces:

- Configuration channel peers cache functionality to the HID forward module. The option allows host tools to retrieve information about change of the connected peripherals without need of complete rediscovery.
- Support for the peers cache in the HID configurator script.
- Configuration channel implementation cleanups.
- Fix for invalid assertion in config event logging implementation.

Jira: NCSDK-17179

### CODE REVIEW

**Comment by MarekPieta** [comment]
Fixed compliance issues.

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 1]
```diff
@@ -997,6 +1023,12 @@ static void disconnect_peripheral(struct hids_peripheral *per)
 	if (per->cfg_chan_rsp) {
 		submit_forward_error_rsp(per, CONFIG_STATUS_WRITE_FAIL);
 	}
+
+	size_t per_id = per - &peripherals[0];
+
+	peripheral_cache[per_id]++;
+	/* A disconnected peripheral is marked with an even number. */
+	__ASSERT_NO_MSG(!(peripheral_cache[per_id] & 1));
```
Comment (kapi-no) [inline applications/nrf_desktop/src/modules/hid_forward.c:1031]: Convert constant `1` to define (e.g. CONNECTED_PEER_DISTINGUISHER or any other name that could hint what it is used for). Also it would be good to define constant as hex `0x01` as it is mainly used in the bitmask operations

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 2]
```diff
@@ -997,6 +1023,12 @@ static void disconnect_peripheral(struct hids_peripheral *per)
 	if (per->cfg_chan_rsp) {
 		submit_forward_error_rsp(per, CONFIG_STATUS_WRITE_FAIL);
 	}
+
+	size_t per_id = per - &peripherals[0];
```
Comment (kapi-no) [inline applications/nrf_desktop/src/modules/hid_forward.c:1027]: Consider defining a simple function for calculating per_id from the per instances (e.g. per_to_per_id). Could it be used also in other places? (If so, i would be great to add a dedicated commit with a small refactor before adding this functionality).

File: scripts/hid_configurator/NrfHidDevice.py [hunk 3]
```diff
@@ -377,10 +390,23 @@ def _read_device_info(dev, recipient):
...
+                                                                   0,
+                                                                   ConfigStatus.GET_PEERS_CACHE,
+                                                                   None)
+        if success:
+            return " ".join("0x{:02x}".format(b) for b in fetched_data)
+        else:
+            return None
+
     @staticmethod
     def _get_connected_peers(dev, recipient):
```
Comment (kapi-no) [inline scripts/hid_configurator/NrfHidDevice.py:406]: why do we attach also peers cache results to this function? Can't we have two standalone functions one for getting connected peers and for getting peers cache?

File: scripts/hid_configurator/NrfHidDevice.py [hunk 4]
```diff
@@ -211,7 +213,18 @@ def open_devices(vid):
                         dir_devs[hwid] = discovered_dev
                         dev_active = True
 
-                peers = NrfHidDevice._get_connected_peers(dev, LOCAL_RECIPIENT)
+                peers, peers_cache = NrfHidDevice._get_connected_peers(dev, LOCAL_RECIPIENT)
```
Comment (kapi-no) [inline scripts/hid_configurator/NrfHidDevice.py:216]: If you separate functions, you can poll peer cache here and then after handling it you can poll for connected peers

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 5]
```diff
@@ -773,35 +778,31 @@ static void handle_config_channel_peers_req(const struct config_event *event)
...
-			if (cur_per < ARRAY_SIZE(peripherals)) {
-				per = &peripherals[cur_per];
+
+			if (per->cfg_chan_id != CFG_CHAN_UNUSED_PEER_ID) {
+				break;
 			}
-		}
 
-		BUILD_ASSERT(sizeof(per->hwid) == HWID_LEN);
+			per = NULL;
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c:793]: It would. Changed. I also renamed `cur_per_idx` to `cur_per` to prevent confusion

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 1 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c:1031]: Changed to `0x01`, thanks for the tip.
I think comment is more meaningful than constant name here (it's difficult to find meaningful name), so I would prefer to keep it as is (also the code is used only in assertions).

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 2 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c:1027]: Currently this is the only place where we use it (there is another function that finds peripheral by configuration channel recipient and it's used in other places). Can we keep it as is?

File: scripts/hid_configurator/NrfHidDevice.py (hunk 3 above)
Comment (MarekPieta) [reply scripts/hid_configurator/NrfHidDevice.py:406]: I prefer to keep it together (from the end-user perspective cache value is useless standalone).  Later on the `_get_connected_peers` could use peers cache internally to avoid rediscovery if possible (currently we do not maintain list of peers so it's not used).

Comment (MarekPieta) [reply scripts/hid_configurator/NrfHidDevice.py:406]: If we add a link between peripherals and dongle (in similar way as it's done in fwupd), we would update it

File: scripts/hid_configurator/NrfHidDevice.py (hunk 4 above)
Comment (MarekPieta) [reply scripts/hid_configurator/NrfHidDevice.py:216]: I would prefer to get peers cache after indexing peers to prevent accidentally triggering assertion failure (the one in config_event logging implementation) on peripherals with older firmware.

Comment (MarekPieta) [reply scripts/hid_configurator/NrfHidDevice.py:216]: Also, as stated in previous comment, later on we could use peers cache only internally in the function.

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 6]
```diff
@@ -813,6 +812,26 @@ static void handle_config_channel_peers_req(const struct config_event *event)
...
+
+		size_t pos = 0;
+		struct config_event *rsp = generate_response(event,
+							     CONFIG_CHANNEL_FETCHED_DATA_MAX_SIZE);
+
+		memcpy(&rsp->dyndata.data[pos], peripheral_cache, sizeof(peripheral_cache));
+		pos += sizeof(peripheral_cache);
+
+		memset(&rsp->dyndata.data[pos], 0x00, CONFIG_CHANNEL_FETCHED_DATA_MAX_SIZE - pos);
+		pos = CONFIG_CHANNEL_FETCHED_DATA_MAX_SIZE;
```
Comment (mkapala-nordic) [inline applications/nrf_desktop/src/modules/hid_forward.c:828]: nit: i think this line is unnecessary as `pos` is not further used here.

**Review by mkapala-nordic — APPROVED** [review]

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 6 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c:828]: To be honest I preferred to update it anyway (it does not cost much and data consistency is ensured :))

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (kapi-no) [reply applications/nrf_desktop/src/modules/hid_forward.c]: oh, yes. i accidentally named in the same way as indexing variable

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 2 above)
Comment (kapi-no) [reply applications/nrf_desktop/src/modules/hid_forward.c:1025]: yes

File: scripts/hid_configurator/NrfHidDevice.py (hunk 3 above)
Comment (kapi-no) [reply scripts/hid_configurator/NrfHidDevice.py:406]: It is not useless as you make many operations in your open_devices function without using the connected peers variable.

And according to cache encoding, you can deduce the number of connected peers without polling the connected peers

Comment (MarekPieta) [reply scripts/hid_configurator/NrfHidDevice.py:406]: Yes, the peers cache can be very useful for the **rediscovery**, but it would not be too useful before initial discovery (keep in mind that current implementation of the HID configurator script does not store any information about peers after script finishes execution so peers are fully rediscovered each time script is called by the user).

The only information provided by reading peers cache before the initial discovery is the number of connected peers (keep in mind that a given peer may also "move" in the peripherals array on dongle side - it may be linked with various index of the peers cache). So that actually with the current implementation, the peers cache is mainly read for debug purposes here (there is no dedicated use-case for it at this point). Still I would prefer to keep the functionality testable in scope of NCS.

Comment (kapi-no) [reply scripts/hid_configurator/NrfHidDevice.py:406]: you can keep it testable, I am just talking about two dedicated API calls instead of having one that returns two values

Comment (MarekPieta) [reply scripts/hid_configurator/NrfHidDevice.py:406]: If I do it, I would need to get peers cache unconditionally. Current implementation allows me to avoid doing it for peripherals (early return on index peers if functionality is not supported). That helps to avoid assertion failures for peripherals with older firmware. Maybe we could keep it like this? IMHO, it's nothing bad in returning 2 values - especially if they are tightly coupled like in our case.

Comment (kapi-no) [reply scripts/hid_configurator/NrfHidDevice.py:406]: ok

**Review by kapi-no — APPROVED** [review]
